libraries/AP_Generator/AP_Generator_IE_2400.cpp

```diff
@@ -0,0 +1,188 @@
+    _output.consumed_mah += _output.current * (now - _state.last_time_ms) * AMS_TO_MAH;
+
+    _state.last_time_ms = now;
+}
+
+// Process characters received and build sentence for IE 2.4kW
+void AP_Generator_IE_2400::decode_latest_term()
+{
+    // Null terminate and move onto the next term
+    _term[_term_offset] = 0;
+    _term_offset = 0;
+    _term_number++;
+
+    switch (_term_number) {
+        case 1:
+            // Float
+            _parsed.tank_bar = atof(_term);
+            break;
+
+        case 2:
+            // Float
+            _parsed.battery_volt = atof(_term);
+            break;
+
+        case 3:
+            // Signed int base 10
+            _parsed.pwr_out = strtol(_term, nullptr, 10);
+            break;
+
+        case 4:
+            // Unsigned int base 10
+            _parsed.spm_pwr = strtoul(_term, nullptr, 10);
+            break;
+
+        case 5:
+            // Signed int base 10
+            _parsed.battery_pwr = strtol(_term, nullptr, 10);
+            break;
+
+        case 6:
+            // Unsigned int base 10
+            _parsed.state = strtoul(_term, nullptr, 10);
+            break;
+
+        case 7:
+            // Unsigned int base 10
+            _parsed.err_code = strtoul(_term, nullptr, 10);
+            // Sentence only declared valid when we have the expected number of terms
+            _sentence_valid = true;
+            break;
+
+        default:
+            // If we get any other term number something has gone wrong
+            _sentence_valid = false;
+            break;
+    }
+}
+
+// Lookup table for IE 2.4 kW fuel cell failsafes.
+const AP_Generator_IE_2400::Lookup_FS_IE24 AP_Generator_IE_2400::lookup_fs_ie24[] = {
+    { ErrorCode::REDUCED_POWER, (1U << 0)},
+    { ErrorCode::SPM_LOST, (1U << 1)},
+    { ErrorCode::PRESSURE_LOW, (1U << 2)},
+    { ErrorCode::BATTERY_LOW, (1U << 3)},
+    { ErrorCode::PRESSURE_ALERT, (1U << 4)},
+    { ErrorCode::START_DENIED, (1U << 5)},
+    { ErrorCode::SYSTEM_CRITICAL, (1U << 6)},
+    { ErrorCode::PRESSURE_CRITICAL, (1U << 7)},
+    { ErrorCode::BATTERY_CRITICAL, (1U << 8)},
+};
+
+// Convert error code from fuel cell to AP failsafe bitmask
+void AP_Generator_IE_2400::set_failsafe_mask(uint32_t err_in)
+{
+    _state.failsafe = 0;
+
+    // Only search failsafe lookup if error code has changed and
+    // we want to ignore MINOR_INTERNAL error
+    if (err_in == _state.last_err_code ||
+        err_in == (uint32_t)ErrorCode::MINOR_INTERNAL) {
+        return;
+    }
+
+    // Convert the error code to the failsafe bitmask for IE_650_800
+    for (auto &entry : lookup_fs_ie24) {
+        if (entry.option == (enum ErrorCode)err_in) {
+            _state.failsafe = entry.bit_mask;
+            break;
+        }
+    }
+}
+
+// Check error codes and populate message with error code
+bool AP_Generator_IE_2400::check_for_err_code(char* msg_txt, uint8_t msg_len, bool changed_check)
```

Comment:
This might be neater factored into two methods and doing without changed_check.

So one method just returns false if there are not errors, the second wraps that one with the additional "has it changed" check before returning zero.